### PR TITLE
Implement Calendar algorithm with included tests

### PR DESCRIPTION
**Pull request by BryanTriana**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -121,6 +127,50 @@ public void everyAttendeeIsConsidered() {
     Assert.assertEquals(expected, actual);
   }
```
Comment (codewithjess) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:129]: How about adding a test where an event has more than one attendee?

something like:
`new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
            Arrays.asList(PERSON_A, PERSON_B, PERSON_C)`

Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:131]: OK I added new tests for events with multiple attendees in latest commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,211 @@
...
+   * @param request the meeting request used to validate the time interval
+   * @return a list containing the available intervals
+   */
+  private static List<TimeRange> getAvailableIntervals(
+      List<TimeRange> busyIntervals, MeetingRequest request) {
+    if (busyIntervals.isEmpty()) {
+      List<TimeRange> availableIntervals = new ArrayList<>();
+      availableIntervals.add(TimeRange.WHOLE_DAY);
+
+      return availableIntervals;
```
Comment (codewithjess) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:117]: can this be simplified to

`return Arrays.asList(TimeRange.WHOLE_DAY);`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,211 @@
...
+   */
+  private static boolean eventConflictsWithRequest(
+      Set<String> eventAttendees, Set<String> requestedAttendees) {
+    for (String attendee : eventAttendees) {
+      if (requestedAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+
+    return false;
```
Comment (codewithjess) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:100]: can this be simplified to

`!Collections.disjoint(eventAttendees, requestedAttendees)`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,211 @@
...
+    List<TimeRange> optionalIntervals = new ArrayList<>();
+    Set<String> optionalAttendees = new HashSet<>(request.getOptionalAttendees());
+
+    List<TimeRange> mandatoryIntervals = new ArrayList<>();
+    Set<String> mandatoryAttendees;
+
+    if (request.getAttendees().isEmpty()) {
+      mandatoryAttendees = new HashSet<>(optionalAttendees);
+    } else {
+      mandatoryAttendees = new HashSet<>(request.getAttendees());
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: can you add some documentation around these parts, not clear why optional treated as mandatory. Maybe we can rename a little bit?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,211 @@
...
+    Collections.sort(mandatoryIntervals, TimeRange.ORDER_BY_START);
+    Collections.sort(optionalIntervals, TimeRange.ORDER_BY_START);
+
+    List<TimeRange> mergedMandatoryIntervals = getMergedIntervals(mandatoryIntervals);
+    List<TimeRange> mergedOptionalIntervals = getMergedIntervals(optionalIntervals);
+
+    List<TimeRange> availableMandatoryIntervals =
+        getAvailableIntervals(mergedMandatoryIntervals, request);
+    List<TimeRange> availableOptionalIntervals =
+        getAvailableIntervals(mergedOptionalIntervals, request);
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Kind of feels like this section can be extracted into a helper function and called twice instead of all this duplication

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK I made this change in latest commit

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK I made this change in latest commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,188 @@
...
+
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:72]: seems like it's possible to make this a helper function instead of needing code duplication?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,188 @@
...
+    return availableIntervals;
+  }
+
+  /**
+   * Finds the available intervals from a list of busy intervals by calculating for their
+   * complement.
+   *
+   * @param busyIntervals the list containing busy non-overlapping intervals
+   * @param request the meeting request used to validate the time interval
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:88]: If this is true, you shouldn't need to pass in a MeetingRequest object, just a TimeRange right? (Always good to prefer providing the most basic info to an API. It keeps the code simpler and makes your function's intention a lot more clear)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,188 @@
...
+      TimeRange currInterval = intervals.get(i);
+
+      if (mergedInterval.end() < currInterval.start()) {
+        mergedIntervals.add(currInterval);
+      } else if (mergedInterval.end() < currInterval.end()) {
+        mergedInterval = TimeRange.fromStartEnd(mergedInterval.start(), currInterval.end(), false);
+
+        mergedIntervals.remove(mergedIntervals.size() - 1);
+        mergedIntervals.add(mergedInterval);
+      }
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:144]: what's protecting your code from having a recently merged interval in a previous iteration having an end() that's > currInterval.end()? (can be a TODO for another pull request)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,188 @@
...
+
+    return mergedIntervals;
+  }
+
+  /**
+   * Merges two interval lists into one given that each merged interval is greater than or equal to
+   * the meeting duration.
+   *
+   * @param intervalsA the first interval list to merge
+   * @param intervalsB the second interval list to merge
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:155]: missing a @param, and same here where you may just need to pass in a TimeRange

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,188 @@
...
+   * duration of the meeting request.
+   *
+   * @param intervals list in which the new interval will be added
+   * @param interval the interval that is checked for validation
+   * @param request the meeting request with the minimum required duration
+   * @return new list of intervals with a new interval added if it was valid
+   */
+  private static List<TimeRange> addIntervalIfValid(
+      List<TimeRange> intervals, TimeRange interval, MeetingRequest request) {
+    List<TimeRange> newIntervals = new ArrayList<>(intervals);
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:193]: Usually we try to avoid unnecessary new object, can this just be:
void addIntervalIfValid(...)?

Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:195]: I could mutate the list inside the function instead of returning a new one but that could also be confusing since I would be modifying one of the arguments given and Java is pass-by-value. Do you think it is a good idea to change the method so that it modifies the list?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:144]: Oh I forgot to add in the JavaDoc for the function that the list is supposed to be sorted based on ascending order for either starting time or ending time of TimeRange. If the list is sorted then this won't happen.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK I added param and changed parameter to accept TimeRange object instead of request

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK I changed request parameter to TimeRange parameter in latest commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK I removed code duplication by adding helper functions in latest commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Yeah I removed code duplication using helper functions in latest commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This part is no longer there in latest commit, I added some more comments to explain the case in which there are no mandatory attendees.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,211 @@
...
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
+/**
+ *  Holds utility method 'query' to find the available times for a meeting
```
[the remark this answers is not in the document]
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK I made this change in latest commit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK this function is not there anymore in latest commit, but I'll keep in mind this preference for future code reviews.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -81,60 +64,61 @@
...
+   * @param meetingDuration the minimum meeting duration
+   * @return list with all non-overlapping available times for the meeting sorted in ascending order
    */
-  private static boolean eventConflictsWithRequest(
-      Set<String> eventAttendees, Set<String> requestedAttendees) {
-    for (String attendee : eventAttendees) {
-      if (requestedAttendees.contains(attendee)) {
-        return true;
+  private static List<TimeRange> getAvailableIntervals(
+      Collection<Event> events, Collection<String> attendees, Long meetingDuration) {
```
Comment (codewithjess) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: `meetingDuration` should include the time unit

Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: OK I added this change in last commit. Will wait for Rodrigo to review again.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -121,6 +129,50 @@ public void everyAttendeeIsConsidered() {
...
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:171]: java style: for parameters that aren't obvious (especially booleans), we usually add an inline comment to specify what it is. So for here:

TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, /* inclusive= */ true)

Comment (BryanTriana) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:171]: OK I added inline comments in last commit.
